### PR TITLE
[HttpFoundation] Adds various return-typed getters to ParameterBag

### LINKED ISSUE #45775: Add various return-typed getters to ParameterBag
### Description

right now, `ParameterBag` has only a few getter options for fetching values:

```
get(): mixed
getAlpha(): string
getAlnum(): string
getDigits(): string
getInt(): int
getBoolean(): bool
```

For those of us using static analyzers like PHPStan, we often find ourselves having to add a bunch of manual type checking code to ensure that the parameter we are fetching is of a specified type.  I see two main problems/annoyances here:

1. In terms of fetching configuration parameters, this often means implementing type checks to ensure that the returned value is indeed the type that we already know it to be.
2. In terms of fetching Request and Query parameters, we have to contend with the fact that forms cast everything to strings, which means fetching a mixed result from `get()`, type checking it as a string, then casting it to the actual type it should be. 
 
The two exclusions here are `getInt()` and `getBoolean()`. Furthermore, `getInt()` is potentially dangerous with it's naked cast, as you may be fetching `5.5` or `"5.5"` and it will return `5` instead of throwing an exception.

### Example

Would the community find it valuable to have many more typed getters along the lines of `getInt()` and `getBoolean()` that return *validated* parameters returned as their proper type? I'm thinking things along the lines of:

```
getArray(): array
getArrayOrNull(): ?array
getBoolean(): bool
getBooleanOrNull(): ?bool
getInt(): int
getIntOrNull(): ?int
getFloat(): float
getFloatOrNull(): ?float
getString(): string
getStringOrNull(): ?string
```

here are some quick examples, including an improvement of `getInt()`:

```php
protected function getInt(string $key, int $default = 0): int
{
    if (is_int($value = $this->get($key, $default))) {
        return $value;
    }
    throw $this->getException("int", $value);
}

protected function getIntOrNull(string $key, int $default = 0): ?int
{
    $value = $this->get($key, $default)
    if (is_int($value) || is_null($value)) {
        return $value;
    }
    throw $this->getException("int|null", $value);
}

protected function getString(string $key, string $default = ""): string
{
    if (is_string($value = $this->get($key, $default))) {
        return $value;
    }
    throw $this->getException("string", $value);
}

protected function getStringOrNull(string $key, string $default = ""): ?string
{
    $value = $this->get($key, $default)
    if (is_string($value) || is_null($value)) {
        return $value;
    }
    throw $this->getException("string|null", $value);
}

/** @param mixed $value */
private function getException(string $expected, $value): TypeException
{
    return new TypeException("expected parameter of type $expected but found " . gettype($value));
}
```

If the Symfony developers think this is a good idea and/or community members want it, I can work on a PR.  I just didn't want to do the work only to find that this discussion has already been had, or there's something I'm missing, or such a feature wouldn't be accepted by Symfony.

And perhaps this logic shouldn't live in `ParameterBag`.  Maybe it warrants a new `Typer` object that can be recycled all over the place.  I currently have this type of logic implemented on a Trait that I include as needed.

### PR DESCRIPTION
**Pull request by mleczakm**
| Q             | A
| ------------- | ---
| Branch?       | 6.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #45775 
| License       | MIT
| Doc PR        | symfony/symfony-docs#...

### CODE REVIEW

File: src/Symfony/Component/HttpFoundation/Tests/ParameterBagTest.php [hunk 1]
```diff
@@ -378,6 +378,30 @@ public function testGetEnumThrowsExceptionWithInvalidValueType()
...
+    {
+        $bag = new ParameterBag(['valid-value' => 'foo']);
+
+        $this->assertSame('foo', $bag->getStringOrNull('valid-value'));
+        $this->assertNull($bag->getStringOrNull('invalid-key'));
+    }
+
+    public function testGetIntOrNull(): void
+    {
+        $bag = new ParameterBag(['valid-value' => 1]);
```
Comment (GromNaN) [inline src/Symfony/Component/HttpFoundation/Tests/ParameterBagTest.php:392]: An assertion with an invalid value is necessary, i.e. with a value that cannot be converted to int. Same for Boolean.

File: src/Symfony/Component/HttpFoundation/Tests/ParameterBagTest.php [hunk 2]
```diff
@@ -378,6 +378,30 @@ public function testGetEnumThrowsExceptionWithInvalidValueType()
...
+
+        $this->assertSame(1, $bag->getIntOrNull('valid-value'));
+        $this->assertNull($bag->getIntOrNull('invalid-key'));
```
Comment (GromNaN) [inline src/Symfony/Component/HttpFoundation/Tests/ParameterBagTest.php]: I would name it `missing-key` for clarity.

File: src/Symfony/Component/HttpFoundation/ParameterBag.php [hunk 3]
```diff
@@ -155,6 +162,13 @@ public function getInt(string $key, int $default = 0): int
         return $this->filter($key, $default, \FILTER_VALIDATE_INT, ['flags' => \FILTER_REQUIRE_SCALAR]) ?: 0;
     }
 
+    public function getIntOrNull(string $key): ?int
+    {
+        return $this->has($key)
```
Comment (GromNaN) [inline src/Symfony/Component/HttpFoundation/ParameterBag.php:167]: Checking if the value is set is not necessary:
`filter_var(null, \FILTER_VALIDATE_INT, \FILTER_NULL_ON_FAILURE)` returns `null`.

**Review by GromNaN — CHANGES_REQUESTED** [review]
Returning `null` as default or in case of invalid value are 2 distinct features. In 7.0, an exception will be thrown when the value is incompatible with the expected return type, one may prefer this behavior over returning `null` indistinctly.

Please check issues reported by fabbot.

File: src/Symfony/Component/HttpFoundation/Tests/ParameterBagTest.php (hunk 1 above)
Comment (mleczakm) [reply src/Symfony/Component/HttpFoundation/Tests/ParameterBagTest.php:392]: I've changed it to using native `getInt` calls etc. so behaviour will be same in both cases.

File: src/Symfony/Component/HttpFoundation/Tests/ParameterBagTest.php (hunk 2 above)
Comment (mleczakm) [reply src/Symfony/Component/HttpFoundation/Tests/ParameterBagTest.php]: Renamed

File: src/Symfony/Component/HttpFoundation/ParameterBag.php (hunk 3 above)
Comment (mleczakm) [reply src/Symfony/Component/HttpFoundation/ParameterBag.php:167]: I've changed it to use `getInt` implementation instead.

**Comment by nicolas-grekas** [comment]
I'm sorry but I don't think adding these methods is beneficial. If one wants to check if an entry exists before, using `has()` works well.

Other issues described in #45775 are solved by #48525 so we can close it IMHO.

**Pull request closed without merging**